core/trino-spi/src/main/java/io/trino/spi/type/Int128Math.java

```diff
@@ -0,0 +1,1331 @@
+            // scale down for 10**rescale is equivalent to scaling down with 5**rescaleFactor first, then with 2**rescaleFactor
+            scaleDownFive(high, low, scaleFactor, result, offset);
+            shiftRight(result[offset], result[offset + 1], scaleFactor, roundUp, result, offset);
+        }
+        else {
+            scaleDownTen(high, low, scaleFactor, result, offset, roundUp);
+        }
+
+        if (negative) {
+            // negateExact not needed since all positive values can be negated without overflow
+            negate(result, offset);
+        }
+    }
+
+    /**
+     * Scale down the value for 5**fiveScale (result := decimal / 5**fiveScale).
+     */
+    private static void scaleDownFive(long high, long low, int fiveScale, long[] result, int offset)
+    {
+        while (true) {
+            int powerFive = Math.min(fiveScale, MAX_POWER_OF_FIVE_INT);
+            fiveScale -= powerFive;
+
+            int divisor = POWERS_OF_FIVES_INT[powerFive];
+            dividePositives(high, low, divisor, result, offset);
+
+            if (fiveScale == 0) {
+                return;
+            }
+
+            high = result[offset];
+            low = result[offset + 1];
+        }
+    }
+
+    /**
+     * Scale down the value for 10**tenScale (this := this / 5**tenScale). This
+     * method rounds-up, eg 44/10=4, 44/10=5.
+     */
+    private static void scaleDownTen(long high, long low, int tenScale, long[] result, int offset, boolean roundUp)
+    {
+        if (high < 0) {
+            throw new IllegalArgumentException("Value must be positive");
+        }
+
+        boolean needsRounding;
+        do {
+            int powerTen = Math.min(tenScale, MAX_POWER_OF_TEN_INT);
+            tenScale -= powerTen;
+
+            int divisor = POWERS_OF_TEN_INT[powerTen];
+            needsRounding = divideCheckRound(high, low, divisor, result, offset);
+
+            high = result[offset];
+            low = result[offset + 1];
+        }
+        while (tenScale > 0);
+
+        if (roundUp && needsRounding) {
+            incrementUnsafe(result);
+        }
+    }
+
+    static void shiftRight(long high, long low, int shift, boolean roundUp, long[] result, int offset)
+    {
+        if (high < 0) {
+            throw new IllegalArgumentException("Value must be positive");
+        }
+
+        if (shift == 0) {
+            return;
+        }
+
+        boolean needsRounding;
+        if (shift < 64) {
+            needsRounding = roundUp && (low & 1L << (shift - 1)) != 0;
+
+            low = (high << 1 << ~shift) | (low >>> shift);
+            high = high >> shift;
+        }
+        else {
+            needsRounding = roundUp && (high & 1L << (shift - 64 - 1)) != 0;
+
+            low = high >> (shift - 64);
+            high = high >> 63;
+        }
+
+        if (needsRounding) {
+            long tmpHigh = incrementHigh(high, low);
```

Comment:
local variables seems redundant

Reply:
They are not. They are there to avoid any inadvertent dependencies between the result of the first call and the second call.